File: Algorithms_Part_1/Algorithm1Sample/Ex2QuickUnion.java
/* *****************************************************************************
 *  Name:
 *  Date:
 *  Description:
 **************************************************************************** */

import edu.princeton.cs.algs4.StdIn;
import edu.princeton.cs.algs4.StdOut;

public class Ex2QuickUnion {

    private int[] id;

    public Ex2QuickUnion(int n) {
        id = new int[n];
        for (int i = 0; i < n; i++)
            id[i] = i;
    }

    private int root(int p) {
        while (id[p] != p)
            p = id[p];
        return p;
    }

    public void union(int p, int q) {
        int pid = root(p);
        int qid = root(q);
        id[pid] = qid;
    }

    public boolean isConnected(int p, int q) {
        return root(p) == root(q);
    }

    @Override
    public String toString() {
        StringBuilder result = new StringBuilder();
        for (int i = 0; i < id.length; i++)
            result.append(id[i]).append(" ");
        return result.toString();
    }

    public static void main(String[] args) {

        int n = StdIn.readInt();
        Ex2QuickUnion qf = new Ex2QuickUnion(n);
        while (!StdIn.isEmpty()) {
            int p = StdIn.readInt();
            int q = StdIn.readInt();

            if (!qf.isConnected(p, q)) {
                qf.union(p, q);
                StdOut.println("Connected(" + p + "," + q + ") : " + qf.toString());
            }
        }

    }
}

File: PythonLearnig/Algorithms_Part_1/Ex2QuickUnion.py
class QuickUnion:

    def __init__(self, n):
        self.data = list(range(n))

    def union(self, p, q):
        pid = self.root(p)
        qid = self.root(q)
        self.data[pid] = qid

    def is_connected(self, p, q):
        return self.root(p) == self.root(q)

    def __str__(self):
        return str(self.data)

    def root(self, p):
        while p != self.data[p]:
            p = self.data[p]
        return p


def test_quick_union():

    n = int(input("Please enter the no of nodes: "))
    m = int(input("Please enter no of unions: "))

    qu = QuickUnion(n)
    for i in range(m):

        p = int(input())
        q = int(input())

        if not qu.is_connected(p, q):
            qu.union(p, q)
            print(f"Connected ({p},{q}) : {qu}")


if __name__ == '__main__':
    test_quick_union()